python/MOCK_sensors.py

```diff
@@ -122,8 +130,148 @@ def publish(self):
+
+            if(self.brokenSensorData is None):
+                randomSeed = get_rosparam_or_default_if_invalid('random_seed', default=None, rosparam_type_cast=str)
+                random.seed(randomSeed)
+
+                self.brokenSensorData = {'brokenSensorID': random.randint(1, 3),
+                                         'brokenSensorAngle_deg': random.random() * 360,
+                                         'brokenSensorSpeed_knots': random.random() * MAX_BROKEN_WINDSPEED_KNOTS}
+
+            self.corrupt_wind_sensor(data, self.brokenSensorData['brokenSensorID'],
+                                     self.brokenSensorData['brokenSensorAngle_deg'],
+                                     self.brokenSensorData['brokenSensorSpeed_knots'])
+
+            rospy.logwarn_throttle(3, "MOCK wind_sensor_{} ".format(self.brokenSensorData['brokenSensorID'])
+                                   + "publishing corrupt data:\n"
+                                   + "angle stuck at {} degrees,".format(self.brokenSensorData['brokenSensorAngle_deg'])
+                                   + " speed stuck at {}".format(self.brokenSensorData['brokenSensorSpeed_knots'])
+                                   + " knots.")
+
+        elif(brokenWindSensorStatus == "rand_const_error"):
+            # This setting chooses a wind sensor and corrupts its data as follows:
+            # the angle published is the original angle plus a random constant theta
+            # the speed published is given by a*(original speed) + b; a and b are random constants
+            # a is within [1 / MAX_WINDSPEED_MULTIPLIER_ERROR, MAX_WINDSPEED_MULTIPLIER_ERROR]
+            # b is within [-MAX_WINDSPEED_ADDER_ERROR_KNOTS, MAX_WINDSPEED_ADDER_ERROR_KNOTS]
+            # theta is within [-MAX_ANGLE_ERROR_DEG, MAX_ANGLE_ERROR_DEG]
+
+            if(self.brokenSensorData is None):
+                randomSeed = get_rosparam_or_default_if_invalid('random_seed', default=None, rosparam_type_cast=str)
+                random.seed(randomSeed)
+
+                if random.random < 0.5:
+                    multiplier = 1 - random.random() * (1 - 1 / (MAX_WINDSPEED_MULTIPLIER_ERROR))
+                else:
+                    multiplier = 1 + random.random() * (MAX_WINDSPEED_MULTIPLIER_ERROR - 1)
+
+                self.brokenSensorData = {'brokenSensorID': random.randint(1, 3),
+                                         'theta': (2*random.random() - 1) * MAX_ANGLE_ERROR_DEG,
+                                         'windspeedMultiplier': multiplier,
+                                         'windspeedAdder': (2*random.random() - 1) * MAX_WINDSPEED_ADDER_ERROR_KNOTS}
+
+            actualAngle, actualSpeed = self.get_wind_measurement(data, self.brokenSensorData['brokenSensorID'])
+
+            newAngle = actualAngle + self.brokenSensorData['theta']
+            newSpeed = max(0, self.brokenSensorData['windspeedMultiplier'] * actualSpeed
```

Comment:
Can absolute value with `fabs()`

Reply:
What was the issue with the built in function abs()? abs() seems to work as expected for floats as far as I can tell

Reply:
The output type of abs is the same as its input, while for fabs it is always a float
```
abs(<int>) = <int>
fabs(<int>) = <float>
abs(<float>) = fabs(<float>) = <float>
```

Thus, if the input data is guaranteed to be a float, we can use `abs()`.

Reply:
Ok that makes sense. The arguments come from expressions containing `numpy.random.ranf()` so I don't think it should matter, but I changed it to `np.fabs()` just in case.